iceoryx2-bb/container/src/slotmap.rs

```diff
@@ -0,0 +1,639 @@
+        idx_to_data_free_list_head: usize,
+        len: usize,
+    }
+
+    impl<T, Ptr: GenericPointer> MetaSlotMap<T, Ptr> {
+        fn next(&self, start: SlotMapKey) -> Option<(SlotMapKey, &T)> {
+            let idx_to_data = &self.idx_to_data;
+
+            for n in start.0..idx_to_data.len() {
+                let data_idx = self.idx_to_data[n];
+                if data_idx != INVALID {
+                    return Some((
+                        SlotMapKey(n),
+                        self.data[data_idx].as_ref().expect(
+                            "By contract, data contains a value when idx_to_data contains a value",
+                        ),
+                    ));
+                }
+            }
+
+            None
+        }
+
+        pub(crate) unsafe fn initialize_data_structures(&mut self) {
+            let capacity = self.capacity_impl();
+            for n in 0..capacity {
+                self.idx_to_data.push_impl(INVALID);
+                self.data.push_impl(None);
+                self.data_next_free_index.push_impl(n);
+
+                let previous = if n == 0 { INVALID } else { n - 1 };
+                let next = if n < capacity - 1 { n + 1 } else { INVALID };
+                self.idx_to_data_free_list
+                    .push_impl(FreeListEntry { previous, next });
+            }
+        }
+
+        pub(crate) unsafe fn iter_impl(&self) -> Iter<T, Ptr> {
+            Iter {
+                slotmap: self,
+                key: SlotMapKey(0),
+            }
+        }
+
+        pub(crate) unsafe fn contains_impl(&self, key: SlotMapKey) -> bool {
+            self.idx_to_data[key.0] != INVALID
+        }
+
+        pub(crate) unsafe fn get_impl(&self, key: SlotMapKey) -> Option<&T> {
+            match self.idx_to_data[key.0] {
+                INVALID => None,
+                n => Some(self.data[n].as_ref().expect(
+                    "data and idx_to_data correspond and this value must be always available.",
+                )),
+            }
+        }
+
+        pub(crate) unsafe fn get_mut_impl(&mut self, key: SlotMapKey) -> Option<&mut T> {
+            match self.idx_to_data[key.0] {
+                INVALID => None,
+                n => Some(self.data[n].as_mut().expect(
+                    "data and idx_to_data correspond and this value must be always available.",
+                )),
+            }
+        }
+
+        unsafe fn acquire_next_free_index(&mut self) -> Option<usize> {
+            if self.idx_to_data_free_list_head == INVALID {
+                return None;
+            }
+
+            let free_idx = self.idx_to_data_free_list_head;
+            let next = self.idx_to_data_free_list[free_idx].next;
+
+            if next != INVALID {
+                self.idx_to_data_free_list[next].previous = INVALID;
+            }
+            self.idx_to_data_free_list_head = next;
+            Some(free_idx)
+        }
+
+        unsafe fn reserve_index(&mut self, idx: usize) {
```

Comment:
This is a bit misleading, since it does not only reserve but also re-use an already used slot. How about `reserve_or_reuse_index`? `claim_index` might also work.